### PR TITLE
Add mkdocs python classy to document classes

### PR DESCRIPTION
**Pull request by itdependsnetworks**
# Closes: Partial #1570 and potentially #3576
# What's Changed

This is POC, it is only a POC, not intended to go in any form as shown. 

This is a POC around building "classy" docs, that show you the nested inheritance of Python Classes. I created a mkdocs plugin to build this out.

The library provides the following features

* The clear definition of the MRO (Method Resolution Order)
* The code in each method
    * The code even when over written, such as when you would like to see `super` being called
    * The code is provided in an ordered (highest is preferred) manner
* The code provided with line numbers from where the code was take from
* The definition of where the code is preferred
* The code is provided in a collapsible block
* Each attribute, and when not preferred shown via a ~~strike-through~~
* Link-able ancestors, when that class is also covered
* Link-able descendants, if applicable 
* Import statement convenience method

ui-view1
ui-view2
ui-view3

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

There is a few challenges, most importantly is it currently requires you to start the application to run. While I have mostly worked around that when "serving the docs", I haven't when building. There are likely other issues, but again, wanted to show rather than tell.

### CODE REVIEW

**Comment by bryanculver** [comment]
I'm curious to know if it will be possible to get this to work when it comes time to publish http://docs.nautobot.com: https://readthedocs.org/projects/nautobot/builds/20160099/

**Comment by Kircheneer** [comment]
As to what I was aiming at with #3756, the docstring plays a major role there as it already contains valuable documentation a lot of the time, but is currently "hidden" in the source code

File: nautobot/extras/plugins/validators.py [hunk 1]
```diff
@@ -40,6 +41,8 @@ def wrap_model_clean_methods():
     """
     Helper function that wraps plugin model validator registered clean methods for all applicable models
     """
+    if os.getenv("NAUTOBOT_NO_DB"):
+        return
     for model in ContentType.objects.filter(FeatureQuery("custom_validators").get_query()):
```
Comment (bryanculver) [inline nautobot/extras/plugins/validators.py:46]: Might be unncesscesary with #3633

**Comment by itdependsnetworks** [comment]
The workflow has been figured out, and is now building and rendering to read the docs: https://docs-landing-page--3614.org.readthedocs.build/projects/core/en/3614/development/classy/views.html

The things pending would be:
* Get agreement to move forward
* Get mkdocs-python-classy on pypi
* Fix anything left not passing in tests

File: .readthedocs.yaml [hunk 2]
```diff
@@ -11,13 +11,17 @@ build:
   os: "ubuntu-20.04"
   tools:
     python: "3.9"
+  commands:
+    - pip install poetry
```
Comment (bryanculver) [inline .readthedocs.yaml:15]: If this lets us ditch `requirements.txt` in factor of something like `poetry install -E docs`... 🔥

File: pyproject.toml [hunk 3]
```diff
@@ -207,6 +207,9 @@ watchdog = "~2.1.9"
 [tool.poetry.scripts]
 nautobot-server = "nautobot.core.cli:main"
 
+[tool.poetry.group.dev.dependencies]
+mkdocs-python-classy = {git = "https://github.com/itdependsnetworks/mkdocs-python-classy.git", rev = "main"}
```
Comment (bryanculver) [inline pyproject.toml:211]: Depending on how Poetry handles this, the biggest blocker is this must be available in PyPI.

1.3 got held up due to Git imports. PyPI doesn't allow it: https://github.com/nautobot/nautobot/pull/1657, https://github.com/nautobot/nautobot/actions/runs/2185417492/jobs/3209161850 (logs gone but this was the failed release).

File: docker/Dockerfile [hunk 4]
```diff
@@ -312,6 +310,9 @@ USER nautobot
 WORKDIR /opt/nautobot
 
 COPY --from=final-dev --chown=nautobot /opt/nautobot/nautobot_config.py /opt/nautobot/nautobot_config.py
+RUN mkdocs build --strict=False -q
```
Comment (glennmatthews) [inline docker/Dockerfile:313]: Why `--strict=False`?

File: docker/Dockerfile [hunk 5]
```diff
@@ -312,6 +310,9 @@ USER nautobot
...
+# Copy in the built docs to bundle with the wheel
+COPY /source/nautobot/project-static/docs /source/nautobot/project-static/docs
```
Comment (glennmatthews) [inline docker/Dockerfile:315]: This doesn't seem right to me?
1) Same source and target paths?
2) Wheel has already been built at this point so this is too late to bundle in, in any case.

File: docker/Dockerfile [hunk 6]
```diff
@@ -191,18 +191,16 @@ FROM dependencies-dev-python as build-docs
 # Copy docs dependencies to build the rendered docs
 COPY mkdocs.yml /source/
 COPY docs /source/docs
+COPY scripts /source/scripts
 
 # Build the rendered docs, this ensures that the docs are in the final image.
-RUN mkdocs build --no-directory-urls
```
Comment (glennmatthews) [inline docker/Dockerfile:196]: Is the entire `build-docs` layer irrelevant now?

File: docker/Dockerfile [hunk 7]
```diff
@@ -191,18 +191,16 @@ FROM dependencies-dev-python as build-docs
...
 
 FROM dependencies as build-nautobot
 # Copy in the Nautobot source code to build the wheel from
 COPY nautobot /source/nautobot
 
-# Copy in the built docs to bundle with the wheel
-COPY --from=build-docs /source/nautobot/project-static/docs /source/nautobot/project-static/docs
 
 # Build the wheel
 RUN poetry build
```
Comment (glennmatthews) [inline docker/Dockerfile:206]: Do we not still need to render the docs before building the wheel?

File: mkdocs.yml [hunk 8]
```diff
@@ -201,6 +217,11 @@ nav:
           - nautobot.apps.ui: "code-reference/nautobot/apps/ui.md"
           - nautobot.apps.urls: "code-reference/nautobot/apps/urls.md"
           - nautobot.apps.views: "code-reference/nautobot/apps/views.md"
```
Comment (glennmatthews) [inline mkdocs.yml:219]: Does the addition of classy make the above `mkdocstrings` output irrelevant/redundant? It seems confusing to me to have two different sets of rendered docs for the same APIs.

**Review by glennmatthews — COMMENTED** [review]
Interesting stuff!

File: docker/Dockerfile (hunk 4 above)
Comment (itdependsnetworks) [reply docker/Dockerfile:313]: Need to fix, I had other issues and was trying to get around it.

File: docker/Dockerfile (hunk 6 above)
Comment (itdependsnetworks) [reply docker/Dockerfile:196]: I don't know... I am not really sure what the layering buys us. I see the docs on top of the docs, but tbh don't understand it all.

File: mkdocs.yml (hunk 8 above)
Comment (itdependsnetworks) [reply mkdocs.yml:219]: I don't think so... The mkdocstrings is to get that code in those files, while this provides attributes and methods within MRO in mind.

**Comment by bryanculver** [comment]
@jathanism didn't close this. It's because the target branch got deleted.

**Comment by jathanism** [comment]
> @jathanism didn't close this. It's because the target branch got deleted.

👀

**Comment by jathanism** [comment]
@itdependsnetworks Can you please update this to point to `develop`?

**Comment by itdependsnetworks** [comment]
I cannot change target or re-open, created new one here: #4202

**Pull request closed without merging**